posts/2018-12-11-hunt-bugs-down-before-they-are-merged.md

```diff
@@ -0,0 +1,362 @@
+them themselves!
+
+## Appendix: merge queues for greater good
+
+Bugs are not the only way to break master, race conditions also apply to
+software process. They may happen when a CI system only runs tests on a branch,
+without rebasing on master first. The following diagram lists three branches,
+`master`, `bob` and `alice`:
+
+<div class="story">
+
+```
+
+bob master alice
+ .    ✓      .   master is green :)
+      |
+ .    +------+   Alice uses `makeRainbow()`
+      |      |
+ .    |      ✓   branch alice is green :)
+      |      |
+ +----+      |   Bob removes `makeRainbow()`
+ |    |      |
+ ✓    |      |   branch `bob` is green :)
+ |    |      |
+ +---->      |   branch "bob" is merged
+      |      |
+      ✓      |   master is green :)
+      |      |
+      <------+   branch "alice" is merged
+      |
+      x          master is red :(
+      |
+      v
+
+```
+
+</div>
+
+Alice forks a branch and used the function `makeRainbow()`. No bug introduced,
+CI is green, everyone's happy. Around the same time Bob decides to do some
+clean up, forks master, and removes the function `makeRainbow()`. From his
+point of view, no one is using it, CI is green, everyone's happy.
+
+Regardless of who merges first -- Alice or Bob -- the end result can't be good:
+Alice's feature makes use of a function that doesn't exist in the codebase
+anymore after Bob's changes. The solution to avoid those issues is to always
+rebase a branch before running CI tests:
+
+<div class="story">
+
+```
+
+bob master alice
+ .    ✓      .   master is green :)
+      |
+ .    +------+   Alice uses `makeRainbow()`
+      |      |
+ .    |      ✓   branch "alice" is green :)
+      |      |
+ +----+      |   Bob removes `makeRainbow()`
+ |    |      |
+ <----+      |   branch "bob" is rebased on master
+ |    |      |
+ ✓    |      |   branch "bob" is green and ready for merge :)
+ |    |      |
+ +---->      |   branch "bob" is merged
+      |      |
+      |      ?   merge is prevented on branch alice :|
+      |      |
+      +------>   branch "alice" is rebased on master
+      |      |
+      |      x   branch "alice" is red :(
+      |      |
+      |      |   Alice writes a new implementation of `makeRainbow()`
+      |      |
+      |      ✓   branch "alice" is green and ready for merge :)
+      |      |
+      <------+   branch "alice" is merged
+      |
+      ✓          master is green :)
+      |
+      v
+
+```
+</div>
+
+The only thing you have to do is to ensure that no branch is merged without
+being strictly on top of master. The simplest solution is to manually keep
+track of the branches that should be merged:
+
+* Whenever someone wants to merge, their branch is enqueued.
+* Exactly one person pops branches from the queue, rebases them, and
+  merges them if CI is happy.
+
+This is a very tedious process and is better automatized by tools like [bors-ng].
+
+[bors-ng]: https://github.com/bors-ng/bors-ng
+
+<div class="story">
+
+### War story
```

Comment:
I wouldn't call this "war story", I'd rather call it "a poor man's merge queue" or "simple and easy PR queues"